### PR TITLE
onedrive: Make link return direct download link

### PR DESCRIPTION
**Pull request by Levi-Hope**
#### What is the purpose of this change?

Make `rclone link onedrive` return direct download link.
Share link of file will be converted to direct link, but folder won't, since Onedrive manages folder download in a different and intransparent way.

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/5381

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] ~~I have added tests for all changes in this PR if appropriate.~~
- [ ] ~~I have added documentation for the changes if appropriate.~~
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/onedrive/onedrive.go
Comment (ivandeex) [inline backend/onedrive/onedrive.go:1511]: You cannot expect that url path will always have 3+ segments.
Sooner or later Microsoft will change something and rclone will panic.
Next, the term "chunk" in rclone is usually related to data bytes, let's use something else.
Also log something loud if all your conditions failed.
Please rewrite this snippet like this:
```
segments := strings.Split(url, "/")
linkType := ""
if len(segments) >= 3 {
  linkType = segments[3]
}
switch {
case f.driveType == driveA && linkType == urlB:
  url = ...
case ...:
  ...
default:
  fs.Logf(..., "Unsupported link type. Cannot convert to downloadable URL")
}
```

Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1511]: Thanks for your review. Suggestions of terms and logs are great. But I'm a little bit confused about the panic handling. I have two concerns:
1. If Microsoft did change its format, this function will return something differently. Instead of waiting for people to notice that something is wrong, wouldn't it be nice to just let it panic so the problem can be noticed and fixed immediately?
2. If we do want to handle some changes, how about rewriting the snippet like this to handle it more properly:
```go
       	url := result.Link.WebURL
        segments := strings.Split(url, "/")
	switch {
	case f.driveType == driveTypePersonal:
		// https://stackoverflow.com/questions/37951114/direct-download-link-to-onedrive-file
		if len(segments) != 5 {
			fs.Logf(f, "Microsoft might change URL format. Cannot convert to downloadable URL")
		} else if segments[3] != "u" {
			enc := base64.StdEncoding.EncodeToString([]byte(url))
			url = "https://api.onedrive.com/v1.0/shares/u!" + enc[:len(enc)-1] + "/root/content"
		}
	case ...:
		...
	default:
		fs.Logf(...)
```
In the case of format changes, the original URL will be used.

Comment (ivandeex) [reply backend/onedrive/onedrive.go]: > 1. ... let it panic so the problem can be noticed...

panic == crash, i.e. wrong program
instead if crashing we must log error

> 2. If we do want to handle some format changes, how about rewriting the snippet like this to handle it more properly:
> Since some case will access up to the 7th segment.
> ```go
> url := result.Link.WebURL
> segments := strings.Split(url, "/")
> switch {
> case f.driveType == driveTypeBusiness:
>     if len(segments) != 8 {
>         fs.Logf(f, "Conversion method is outdated. Cannot convert to downloadable URL")
>     } else if segments[3] != ":f:" {
>         url = strings.Join(segments[0:3], "/") + "/" + segments[5] + "/" + segments[6] +
>             "/_layouts/15/download.aspx?share=" + segments[7]
>     }

before accessing segment #3 this fragment checks that no. of segments is 8, so OK

> case ...:
>     ...
> default:
>     fs.Logf(...)
> ```
> In the case of format changes, the original URL will be used.

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1524]: Generally I think [early returning](https://danp.net/posts/reducing-go-nesting/) is preferred when there's something failing and we don't have anything else to do afterwards. It helps with the readability. (The same applies to others)

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1526]:
```suggestion
			url = strings.Join(segments[:3], "/") + "/" + segments[5] + "/" + segments[6] +
```

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1533]:
```suggestion
			url = strings.Join(segments[:3], "/")
```

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1545]: We shouldn't need the default clause here as we've exhausted the possibilities for `f.driveType`.

**Review by Cnly — CHANGES_REQUESTED** [review]
Thank you for your contribution @Levi-Hope! I have the following comments regarding this PR, some of which may be more important but others are not really so :)

PS: If possible, please don't force push as it makes reviews, discussions, etc. harder to follow.

File: backend/onedrive/onedrive.go
Comment (ivandeex) [reply backend/onedrive/onedrive.go:1513]: Agree.
It's perfectly fine to write
```
switch {
case a==123 && b==456:
...
}
```

But
```
switch {
case a==123:
...
case a==456:
...
}
```

should be simplified into
```
switch a {
case 123:
...
case 456:
...
}
```

File: backend/onedrive/onedrive.go
Comment (Cnly) [reply backend/onedrive/onedrive.go:1516]: @ivandeex I think what you've said is reasonable. In that case Notice is fine for me. Also Notice is a lower level than Warn and can be suppressed with `-q` so I prefer it over Warn which can't be silenced.

I took a closer look and found that the docs of `Logf` says it uses the Info level (which has an `Infof` equivalent) but in fact it uses Notice which has no corresponding level-specific function. That said, the original `fs.Logf(f, cnvFailMsg)` call already looks perfect and needs no changes. Sorry for that!

I'll fix the docs in a while.

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1513]: Sure, this does get better.

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1511]: OK. I'm fine with it.

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1519]: Yep, I will go check with this.

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1524]: Sounds like a good idea.

**Comment by Cnly** [comment]
@Levi-Hope thank you for taking care of the comments. If you've resolved them, could you please also click on "Resolve conversation"?

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1519]:
```suggestion
		// Method: https://stackoverflow.com/questions/37951114/direct-download-link-to-onedrive-file
```

In Go, `//` line comments are [the norm](https://golang.org/doc/effective_go#commentary). And this can also be a one-line comment. (also applies to others)

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1529]: If I understand this correctly, you're just stripping off the last character from the encoded string unconditionally? Does this work when the last character is not a padding character, or if there are two padding characters at the end? (Please feel free to correct me.)

**Comment by Levi-Hope** [comment]
> @Levi-Hope thank you for taking care of the comments. If you've resolved them, could you please also click on "Resolve conversation"?

Done.

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1560]: Maybe it'll be better to also put an upper bound here, just to be safe.

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1572]: If this path is taken, is the returned url usable? If not, we should probably just return the original one.

**Review by Cnly — CHANGES_REQUESTED** [review]
Thanks :) We've got just a few things to discuss and do before we can get this merged.

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1519]: KK

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1529]: Well, actually the Microsoft API works even with the padding character included. But sure, let's strictly follow the format.

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1560]: KK

File: backend/onedrive/onedrive.go
Comment (Levi-Hope) [reply backend/onedrive/onedrive.go:1572]: KK

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1545]:
```suggestion
		//   https://{tenant}.sharepoint.com/:t:/s/{site_name}/{Opaque_String}
```

URLs for Sharepoint don't contain `-my`. (Same applies to others)

File: backend/onedrive/onedrive.go
Comment (Cnly) [inline backend/onedrive/onedrive.go:1512]:
```suggestion
		fs.Debugf(nil, "Can't convert share link for folder to direct link - returning the link as is")
```

IMO we're not expecting share links for folders to be convertible so we don't need to disturb the user

**Review by Cnly — APPROVED** [review]
LGTM in general! I'm still running a few tests on my side now, and we'll be good to go once they're done! :D Meanwhile I just have some remaining nitpicking comments...

**Comment by Cnly** [comment]
It just occurred to me that we should probably be conservative and refrain from blindly attempting the conversion for regional versions. I'll merge once the checks succeed. Thank you again! :)

**Comment by Levi-Hope** [comment]
Great thanks to you guys, nice reviews.

**Pull request merged**
